odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java

```diff
@@ -0,0 +1,344 @@
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector.PrimitiveCategory;
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.BinaryObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.BooleanObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.ByteObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.DateObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.DoubleObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.FloatObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.HiveCharObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.HiveDecimalObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.HiveVarcharObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.IntObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.LongObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.ShortObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.TimestampObjectInspector;
+
+/**
+ * Only for odps 2.0
+ * @author: Jon (wangzhong.zw@alibaba-inc.com)
+ */
+public class OdpsDataTransferUDTF extends GenericUDTF {
+  private ObjectInspector[] objectInspectors;
+  private Odps odps;
+  private TableTunnel tunnel;
+  private UploadSession uploadSession;
+  private RecordWriter recordWriter;
+  private String currentOdpsTableName = "";
+  private String currentOdpsPartitionSpec = "";
+
+  public OdpsDataTransferUDTF() throws IOException {
+    OdpsConfig odpsConfig = new OdpsConfig("odps_config.properties");
+    AliyunAccount account = new AliyunAccount(odpsConfig.getAccessId(), odpsConfig.getAccessKey());
+    this.odps = new Odps(account);
+    this.odps.setDefaultProject(odpsConfig.getProjectName());
+    this.odps.setEndpoint(odpsConfig.getOdpsEndpoint());
+    this.tunnel = new TableTunnel(odps);
+    if (odpsConfig.getTunnelEndpoint() != null) {
+      this.tunnel.setEndpoint(odpsConfig.getTunnelEndpoint());
+    }
+  }
+
+  @Override
+  public StructObjectInspector initialize(ObjectInspector[] args) throws UDFArgumentException {
+    this.objectInspectors = args;
+    // This UDTF doesn't output anything
+    return ObjectInspectorFactory.getStandardStructObjectInspector(new ArrayList<String>(),
+        new ArrayList<ObjectInspector>());
+  }
+
+  @Override
+  public void process(Object[] args) throws HiveException {
+    try {
+      StringObjectInspector soi0 = (StringObjectInspector) objectInspectors[0];
```

Comment:
The table name & columns & partition columns are not likely to change during the life time of the UDTF, so we can read them only once to reduce overhead